proof/crefine/lib/Boolean_C.thy

```diff
@@ -0,0 +1,138 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2014, General Dynamics C4 Systems
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(* Definitions and rules for C boolean constants parsed from the kernel *)
+
+theory Boolean_C
+imports
+  "CSpec.KernelInc_C"
+begin
+
+lemma true_and_1[simp]:
+  "true && 1 = true"
+  "signed true && 1 = signed true"
+  "unsigned true && 1 = unsigned true"
+  by (simp add: true_def)+
+
+lemma to_bool_true[simp]:
+  "to_bool true"
+  "to_bool (signed true)"
+  "to_bool (unsigned true)"
+  by (simp add: to_bool_def true_def)+
+
+lemma true_equals_simps[simp]:
+  "true = 1 \<longleftrightarrow> True"
+  "signed true = 1 \<longleftrightarrow> True"
+  "unsigned true = 1 \<longleftrightarrow> True"
+  "true = 0 \<longleftrightarrow> False"
+  "signed true = 0 \<longleftrightarrow> False"
+  "unsigned true = 0 \<longleftrightarrow> False"
```

Comment:
Not too sure about these as `[simp]` rules, because the conclusion doesn't have much the simplifier can match on. What breaks when you leave them out?

Reply:
These being `[simp]` rules was basically the starting point of this PR. The original purpose of this work was to make it so that simple goals involving `true` and `false` are handled completely automatically, so that people who don't know about them don't need to spend time figuring out that they are special. (To be even more precise, I was one of those people recently and spent too long trying to work out why a goal wasn't being solved before realising that I needed to give it `false_def`.)

I'm waiting for things to rebuild to confirm, but I would expect that putting them in the set as needed will require reverting everywhere that I've removed `true_def` and `false_def` and replacing them with `true_equals_simps` and `false_equals_simps`.

Reply:
Sorry, all good, I once again misread `longleftrightarrow` for `longrightarrow`. Equivalences make complete sense here.